portfolio/src/test/java/com/google/sps/CommentsServletPostTest.java

```diff
@@ -0,0 +1,189 @@
+package com.google.sps;
+
+import com.google.appengine.api.datastore.*;
+import com.google.sps.data.Comments;
+import com.google.sps.servlets.CommentsServlet;
+import org.junit.Test;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Date;
+import java.util.List;
+
+import static org.mockito.Mockito.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.*;
+
+public class CommentsServletPostTest extends ServletTest {
+    private final String commentsPage = "/comments.html";
+    private final String commentText = "text";
+    private final Instant instant = Instant.ofEpochMilli(1234567890L);
+
+    private void doPostRequest(HttpServletRequest request, HttpServletResponse response, Instant instant) throws IOException {
+        CommentsServlet servlet = new CommentsServlet();
+        servlet.init();
+        servlet.setMockInstant(instant);
+        servlet.doPost(request, response);
+    }
+
+    @Test
+    public void testAddComment() throws IOException {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+        when(request.getParameter("action")).thenReturn("add");
+        when(request.getParameter(Comments.TEXT)).thenReturn(commentText);
+
+        doPostRequest(request, response, instant);
+        verify(response).sendRedirect(commentsPage);
+
+        Query query = new Query(Comments.COMMENT_ENTITY_KIND);
+        List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+
+        Entity expected = new Entity(Comments.COMMENT_ENTITY_KIND);
+        expected.setProperty(Comments.DATE, Date.from(instant));
+        expected.setProperty(Comments.TEXT, commentText);
+        expected.setProperty(Comments.RATING, 0L);
+
+        assertThat(results.stream().map(PropertyContainer::getProperties).toArray(),
+                arrayContaining(expected.getProperties()));
```

Comment:
It's better to keep the asserts simple. If you're checking `results.stream().map(PropertyContainer::getProperties).toArray()`, first extract it to a variable with a meaningful name to make it clear what you're testing.

Also, you can use https://truth.dev/ to write more fluent assertions. Up to you.

Same applies to other tests.